docs/devel/owners.md

```diff
@@ -0,0 +1,128 @@
+
+After a PR is submitted, the automated kubernetes PR robot will append a message to the PR indicating the owners
+that are required for the PR to be submitted.
+
+Subsequently, a user can also request the approval message from the robot by writing:
+
+```
+@k8s-bot approvers
+```
+
+into a comment.
+
+In either case, the automation replies with an annotation that indicates
+the owners required to approve.  The annotation is a comment that is applied to the PR.
+This comment will say:
+
+```
+Approval is required from <owner-a> OR <owner-b>, AND <owner-c> OR <owner-d>, AND ...
+```
+
+The set of required authors is drawn from the OWNERS files in the repository (see below).  For each file
+there should be multiple different OWNERS, these owners are listed in the `OR` clause(s). Because
+it is possible that a PR may cover different directories, with disjoint sets of OWNERS, a PR may require
+approval from more than one person, this is where the `AND` clauses come from.
+
+### Step Two: A PR is LGTM'd
+
+Once a PR is reviewed and LGTM'd it is eligible for submission.  However, for it to be submitted
+an owner for all of the files changed in the PR have to 'approve' the PR.  A user is an owner for a
+file if they are included in the OWNERS hierarchy (see below) for that file.
+
+Owner approval comes in two forms:
+
+   * An owner adds a comment to the PR saying "I approve"
+   * An owner is the original author of the PR
+
+In the case of a comment based approval, the same rules as for the 'lgtm' label apply.  If the PR is
+changed by pushing new commits to the PR, the previous approval is invalidated, and the owner(s) must
+approve again.  Because of this is recommended that PR authors squash their PRs prior to getting approval
+from owners.
+
+### Step Three: A PR is merged
+
+Once a PR is LGTM'd and all owners have approved, it is eligible for merge.  The merge bot takes care of
+the actual merging.
+
+## Design details
+
+We need to build new features into the existing github munger in order to accomplish this.  Additionally
+we need to add owners files to the repository.
+
+### Approval Munger
+
+We need to add a munger that adds comments to PRs indicating whose approval they require.  This munger will
+look for PRs that do not have approvers already present in the comments, or where approvers have been
+requested, and add an appropriate comment to the PR.
+
+
+### Status Munger
+
+GitHub has a [status api](https://developer.github.com/v3/repos/statuses/), we will add a status munger that pushes a status onto a PR of approval status.  This status will only be approved if the relevant
+approvers have approved the PR.
+
+### Requiring approval status
+
+Github has the ability to [require status checks prior to merging](https://help.github.com/articles/enabling-required-status-checks/)
+
+Once we have the status check munger described above implemented, we will add this required status check
+to our main branch as well as any release branches.
```

Comment:
Just to put numbers to the discussion. `git log --merges --format="%an" v1.1.0..HEAD | sort | uniq -c` This is all of the people who clicked the big green button between 1.1.0 and HEAD. The people at the bottom are the 'owners' who I assume would be exempt from this enforcement.

I'm not sure what these are telling us. But maybe it is telling us we need to start the social engineering now   :)

```
     12 Abhi Shah
     55 Alex Robinson
      3 CJ Cullen
      2 David Eads
      1 Derek Carr
      2 Dr. Stefan Schimanski
      5 Fabio Yeon
      4 Isaac Hollander McCreery
      4 Janet Kuo
      1 Jeff Grafton
     19 Jeff Lowdermilk
     14 Jerzy Szczepkowski
      3 Karl Isenberg
      3 krousey
      9 Marcin Wielgus
     46 Marek Grabowski
      8 Nikhil Jindal
      3 Paul Morie
      2 Phillip Wittrock
     26 Piotr Szczesniak
      4 Prashanth B
      2 Quinton Hoole
      9 Robert Bailey
      9 Saad Ali
      6 Vincenzo D'Amore
     21 Wojciech Tyczynski
      1 Yu-Ju Hong
      -----------------
    274 Total

    671 k8s-merge-robot

      1 * Alex Mohr
     42 * Brendan Burns
      7 * Brian Grant
     16 * Daniel Smith
      5 * David Oppenheimer
     53 * Dawn Chen
      9 * Eric Tune
     22 * Filip Grzadkowski
     23 * Mike Danese
     11 * Tim Hockin
      2 * Zach Loafman
      ------------------
    191 Total
```